assemblyscript/assembly/convert_image/host/as-wasi-nn.ts

```diff
@@ -0,0 +1,241 @@
+     * @returns the output tensor
+     */
+    getOutput(index: u32, outputBuffer: Array<u8>): Tensor {
+        let maxBufferLength = outputBuffer.length;
+        let bytesWritten: u32 = changetype<u32>(memory.data(4));
+        let resultCode = wasi_ephemeral_nn.get_output(load<u32>(this.pointer), index,
+            changetype<u32>(getArrayPtr(outputBuffer)),
+            maxBufferLength,
+            bytesWritten);
+        if (resultCode != 0) {
+            throw new WasiNnError("Unable to get output tensor", resultCode);
+        }
+        return new Tensor([maxBufferLength], TensorType.u8, outputBuffer);
+    }
+}
+
+/**
+ * Contains the data passed to inference.
+ */
+export class Tensor {
+    constructor(public dimensions: u32[], public type: TensorType, public data: u8[]) { }
+
+    /**
+     * @returns a pointer to a structure describing the tensor; this conforms to the WITX `$tensor`
+     * description as well as the ABI expected by Wasmtime.
+     * @see https://github.com/WebAssembly/wasi-nn/blob/master/phases/ephemeral/witx/wasi_ephemeral_nn.witx#L56
+     * @see https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasi-nn/examples/wasi-nn-rust-bindings/src/generated.rs#L19-L35
+    */
+    asPointer(): u32 {
+        let struct: u32[] = [getArrayPtr(this.dimensions), this.dimensions.length, this.type,
+        getArrayPtr(this.data), this.data.length];
+        return getArrayPtr(struct);
+        // May need to pin `struct` so it is not garbage-collected? (See
+        // https://www.assemblyscript.org/garbage-collection.html#garbage-collection). Perhaps not,
+        // as long as we use the stub runtime (see
+        // https://www.assemblyscript.org/garbage-collection.html#stub-runtime).
+    }
+
+    /**
+     * Convert data to an `ArrayBuffer` for using data views.
+     * @returns an ArrayBuffer with a copy of the bytes in `this.data`
+     */
+    toArrayBuffer(): ArrayBuffer {
+        const buffer = new ArrayBuffer(this.data.length);
+        // TODO figure out why we cannot use `forEach` here.
+        for (let i = 0; i < this.data.length; i++) {
+            store<u8>(changetype<u32>(buffer) + i, this.data[i])
+        }
+        return buffer;
+    }
+}
+
+/**
+ * Convenience function that converts an image to a tensor
+ *
+ * @param path Path to the image
+ * @param width Width of the tensor
+ * @param height Height of the tensor
+ * @param precision precision of the tensor
+ * @returns u8[] containing the tensor if successful. Empty array if it fails
+ */
+export function convert_image(path: string, width: u32, height: u32, precision: TensorType): u8[] {
+
+    let bytes = 1;
+
+    if (precision === TensorType.f32 || precision === TensorType.i32) {
+        bytes = 4;
+    } else if (precision === TensorType.f16) {
+        bytes = 2;
+    }
+
+    // The 3 is because we have Red, Green, and Blue in each pixel.
+    let out_buffer = new Array<u8>(width * height * 3 * bytes).fill(0);
```

Comment:
```suggestion
    let out_buffer = new Array<u8>(width * height * 3 * bytes);
```
In AS you can skip filling array to zeros

Reply:
Good to know, thanks.